File: ios/Ribbon/Services/Auth0Service.swift
import AuthenticationServices
import CommonCrypto
import Foundation
import UIKit

/// Result of an Auth0 authentication ceremony.
public struct Auth0User: Sendable {
    public let idToken: String
    public let userUUID: UUID
    public let email: String?
    public let refreshToken: String

    public init(idToken: String, userUUID: UUID, email: String?, refreshToken: String = "") {
        self.idToken = idToken
        self.userUUID = userUUID
        self.email = email
        self.refreshToken = refreshToken
    }
}

public enum Auth0Error: LocalizedError, Sendable {
    case notConfigured
    case cancelled
    case invalidCallback
    case tokenExchangeFailed(String)
    case malformedToken

    public var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "Auth0 domain or client ID is not configured."
        case .cancelled:
            return "Authentication was cancelled."
        case .invalidCallback:
            return "Invalid callback URL received from Auth0."
        case .tokenExchangeFailed(let details):
            return "Failed to exchange authorization code: \(details)"
        case .malformedToken:
            return "Malformed ID token."
        }
    }
}

@MainActor
private final class WebAuthContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
    private let anchor: ASPresentationAnchor

    init(anchor: ASPresentationAnchor) {
        self.anchor = anchor
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        anchor
    }
}

/// Service managing Auth0 Universal Login authentication, PKCE flow,
/// and token parsing for Supabase integration.
public enum Auth0Service {
    public static let namespaceDNS = "6ba7b810-9dad-11d1-80b4-00c04fd430c8"

    /// Performs the Auth0 Universal Login flow using ASWebAuthenticationSession
    /// with PKCE (RFC 7636). Returns authenticated credentials.
    @MainActor
    public static func login(anchor: ASPresentationAnchor) async throws -> Auth0User {
        guard Auth0Config.isConfigured else {
            throw Auth0Error.notConfigured
        }

        let domain = Auth0Config.domain.trimmingCharacters(in: .whitespacesAndNewlines)
        let clientId = Auth0Config.clientId.trimmingCharacters(in: .whitespacesAndNewlines)
        let scheme = Auth0Config.scheme.trimmingCharacters(in: .whitespacesAndNewlines)
        let bundleId = Bundle.main.bundleIdentifier ?? "bible.ribbon.app"
        let redirectURI = "\(scheme)://\(domain)/ios/\(bundleId)/callback"

        let (verifier, challenge) = generatePKCE()
        let state = UUID().uuidString.lowercased()

        var components = URLComponents(string: "https://\(domain)/authorize")
        components?.queryItems = [
            URLQueryItem(name: "client_id", value: clientId),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "scope", value: "openid profile email offline_access"),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256"),
            URLQueryItem(name: "state", value: state)
        ]

        guard let authURL = components?.url else {
            throw Auth0Error.notConfigured
        }

        let contextProvider = WebAuthContextProvider(anchor: anchor)

        let callbackURL: URL = try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: authURL,
                callbackURLScheme: scheme
            ) { callbackURL, error in
                if let error = error as? ASWebAuthenticationSessionError, error.code == .canceledLogin {
                    continuation.resume(throwing: Auth0Error.cancelled)
                } else if let error {
                    continuation.resume(throwing: error)
                } else if let callbackURL {
                    continuation.resume(returning: callbackURL)
                } else {
                    continuation.resume(throwing: Auth0Error.cancelled)
                }
            }
            session.presentationContextProvider = contextProvider
            session.prefersEphemeralWebBrowserSession = false
            session.start()
        }

        guard let urlComponents = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false),
              let queryItems = urlComponents.queryItems else {
            throw Auth0Error.invalidCallback
        }

        if let errorParam = queryItems.first(where: { $0.name == "error" })?.value {
            let desc = queryItems.first(where: { $0.name == "error_description" })?.value ?? errorParam
            throw Auth0Error.tokenExchangeFailed(desc)
        }

        guard let code = queryItems.first(where: { $0.name == "code" })?.value,
              let returnedState = queryItems.first(where: { $0.name == "state" })?.value,
              returnedState == state else {
            throw Auth0Error.invalidCallback
        }

        // Exchange authorization code for tokens
        guard let tokenURL = URL(string: "https://\(domain)/oauth/token") else {
            throw Auth0Error.notConfigured
        }

        var request = URLRequest(url: tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let bodyDict: [String: String] = [
            "grant_type": "authorization_code",
            "client_id": clientId,
            "code_verifier": verifier,
            "code": code,
            "redirect_uri": redirectURI
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: bodyDict)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            let errorText = String(data: data, encoding: .utf8) ?? "HTTP status \((response as? HTTPURLResponse)?.statusCode ?? 0)"
            throw Auth0Error.tokenExchangeFailed(errorText)
        }

        guard let tokenJSON = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let idToken = tokenJSON["id_token"] as? String else {
            throw Auth0Error.tokenExchangeFailed("Missing id_token in token response")
        }

        let refreshToken = tokenJSON["refresh_token"] as? String ?? ""
        let (userUUID, email) = try parseIdToken(idToken)

        return Auth0User(
            idToken: idToken,
            userUUID: userUUID,
            email: email,
            refreshToken: refreshToken
        )
    }

    /// Refreshes the Auth0 session using a refresh token, returning a fresh ID token and new refresh token.
    public static func refreshTokens(refreshToken: String) async throws -> (idToken: String, refreshToken: String) {
        guard Auth0Config.isConfigured else { throw Auth0Error.notConfigured }

        let domain = Auth0Config.domain.trimmingCharacters(in: .whitespacesAndNewlines)
        let clientId = Auth0Config.clientId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let tokenURL = URL(string: "https://\(domain)/oauth/token") else {
            throw Auth0Error.notConfigured
        }

        var request = URLRequest(url: tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let bodyDict: [String: String] = [
            "grant_type": "refresh_token",
            "client_id": clientId,
            "refresh_token": refreshToken
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: bodyDict)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            let errorText = String(data: data, encoding: .utf8) ?? "HTTP status \((response as? HTTPURLResponse)?.statusCode ?? 0)"
            throw Auth0Error.tokenExchangeFailed(errorText)
        }

        guard let tokenJSON = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let idToken = tokenJSON["id_token"] as? String else {
            throw Auth0Error.tokenExchangeFailed("Missing id_token in refresh response")
        }

        let newRefreshToken = tokenJSON["refresh_token"] as? String ?? refreshToken
        return (idToken, newRefreshToken)
    }

    /// Parses an Auth0 ID token (JWT) without signature verification (which is verified by Supabase).
    /// Extracts `user_uuid` or calculates a deterministic RFC 4122 Version 5 UUID from `sub`.
    public static func parseIdToken(_ token: String) throws -> (userUUID: UUID, email: String?) {
        let parts = token.components(separatedBy: ".")
        guard parts.count >= 2 else { throw Auth0Error.malformedToken }

        var base64 = parts[1]
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while base64.count % 4 != 0 {
            base64.append("=")
        }

        guard let data = Data(base64Encoded: base64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw Auth0Error.malformedToken
        }

        let sub = json["sub"] as? String ?? ""
        let email = json["email"] as? String

        let userUUID: UUID
        if let customUUID = json["user_uuid"] as? String, let parsed = UUID(uuidString: customUUID) {
            userUUID = parsed
        } else {
            userUUID = computeUuidV5(name: sub)
        }

        return (userUUID, email)
    }

    /// Generates a deterministic RFC 4122 Version 5 UUID using SHA-1 and the standard DNS namespace.
    public static func computeUuidV5(name: String, namespace: String = namespaceDNS) -> UUID {
        guard let ns = UUID(uuidString: namespace) else { return UUID() }
        var nsBytes = ns.uuid
        var data = Data(bytes: &nsBytes, count: 16)
        if let nameData = name.data(using: .utf8) {
            data.append(nameData)
        }
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA1(buffer.baseAddress, CC_LONG(buffer.count), &digest)
        }
        digest[6] = (digest[6] & 0x0F) | 0x50
        digest[8] = (digest[8] & 0x3F) | 0x80

        let uuidTuple: uuid_t = (
            digest[0], digest[1], digest[2], digest[3],
            digest[4], digest[5], digest[6], digest[7],
            digest[8], digest[9], digest[10], digest[11],
            digest[12], digest[13], digest[14], digest[15]
        )
        return UUID(uuid: uuidTuple)
    }

    private static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .trimmingCharacters(in: CharacterSet(charactersIn: "="))
    }

    private static func generatePKCE() -> (verifier: String, challenge: String) {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        let verifier = base64URLEncode(Data(bytes))

        let verifierData = verifier.data(using: .utf8) ?? Data()
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        verifierData.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(buffer.count), &digest)
        }
        let challenge = base64URLEncode(Data(digest))
        return (verifier, challenge)
    }
}

File: android/app/src/main/kotlin/app/readribbon/services/Auth0Service.kt
package app.readribbon.services

import android.app.Activity
import app.readribbon.data.Auth0Config
import com.auth0.android.Auth0
import com.auth0.android.authentication.AuthenticationException
import com.auth0.android.callback.Callback
import com.auth0.android.provider.WebAuthProvider
import com.auth0.android.result.Credentials
import java.security.MessageDigest
import java.util.Base64
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.uuid.Uuid
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Result of an Auth0 credentials exchange.
 */
data class Auth0User(
    val idToken: String,
    val userUuid: Uuid,
    val email: String?,
    val refreshToken: String = "",
)

sealed class Auth0Exception(message: String) : Exception(message) {
    data object NotConfigured : Auth0Exception("Auth0 domain or client ID is not configured")
    data object Cancelled : Auth0Exception("Auth0 login was cancelled by user")
    class Failed(val details: String) : Auth0Exception(details)
}

object Auth0Service {
    private const val NAMESPACE_DNS = "6ba7b810-9dad-11d1-80b4-00c04fd430c8"

    /**
     * Launch the Auth0 Universal Login flow using Android Custom Tabs.
     */
    suspend fun login(activity: Activity): Auth0User {
        if (!Auth0Config.isConfigured) {
            throw Auth0Exception.NotConfigured
        }

        val account = Auth0.getInstance(Auth0Config.CLIENT_ID, Auth0Config.DOMAIN)

        val credentials = suspendCancellableCoroutine<Credentials> { continuation ->
            WebAuthProvider.login(account)
                .withScheme(Auth0Config.SCHEME)
                .withScope("openid profile email offline_access")
                .start(activity, object : Callback<Credentials, AuthenticationException> {
                    override fun onSuccess(result: Credentials) {
                        continuation.resume(result)
                    }

                    override fun onFailure(error: AuthenticationException) {
                        if (error.isCanceled) {
                            continuation.resumeWithException(Auth0Exception.Cancelled)
                        } else {
                            continuation.resumeWithException(
                                Auth0Exception.Failed(error.getDescription() ?: error.message ?: "Authentication failed")
                            )
                        }
                    }
                })
        }

        val idToken = credentials.idToken
        val parsed = parseIdToken(idToken)
        return Auth0User(
            idToken = idToken,
            userUuid = parsed.userUuid,
            email = parsed.email,
            refreshToken = credentials.refreshToken ?: "",
        )
    }

    /**
     * Parses an Auth0-issued ID token (JWT) without remote verification (since the token
     * is passed directly to Supabase which validates the signature via OIDC JWKS).
     * Extracts user_uuid (or computes deterministic UUIDv5 from 'sub'), and email.
     */
    fun parseIdToken(idToken: String): Auth0User {
        val parts = idToken.split(".")
        if (parts.size < 2) {
            throw IllegalArgumentException("Malformed JWT: expected at least 2 segments")
        }

        val payloadBase64 = parts[1]
        val decodedBytes = Base64.getUrlDecoder().decode(padBase64(payloadBase64))
        val json = Json.parseToJsonElement(String(decodedBytes, Charsets.UTF_8)).jsonObject

        val sub = json["sub"]?.jsonPrimitive?.contentOrNull ?: ""
        val email = json["email"]?.jsonPrimitive?.contentOrNull

        // Priority 1: user_uuid injected by our Auth0 Post-Login Action
        val customUuid = json["user_uuid"]?.jsonPrimitive?.contentOrNull
        val userUuid = if (!customUuid.isNullOrBlank()) {
            runCatching { Uuid.parse(customUuid) }.getOrNull() ?: computeUuidV5(sub)
        } else {
            computeUuidV5(sub)
        }

        return Auth0User(
            idToken = idToken,
            userUuid = userUuid,
            email = email,
        )
    }

    private fun padBase64(input: String): String {
        val remainder = input.length % 4
        return if (remainder > 0) input + "=".repeat(4 - remainder) else input
    }

    /**
     * Generates a deterministic RFC 4122 Version 5 UUID using SHA-1 and the standard DNS namespace,
     * identical to PostgreSQL's extensions.uuid_generate_v5 and Node's uuidv5.
     */
    fun computeUuidV5(name: String, namespace: String = NAMESPACE_DNS): Uuid {
        val cleanNs = namespace.replace("-", "")
        val nsBytes = ByteArray(16) { i ->
            cleanNs.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
        val nameBytes = name.toByteArray(Charsets.UTF_8)
        val digest = MessageDigest.getInstance("SHA-1")
        digest.update(nsBytes)
        digest.update(nameBytes)
        val hash = digest.digest()

        // Set version 5 (bits 4-7 of byte 6 = 0101)
        hash[6] = ((hash[6].toInt() and 0x0f) or 0x50).toByte()
        // Set variant to RFC 4122 (bits 6-7 of byte 8 = 10)
        hash[8] = ((hash[8].toInt() and 0x3f) or 0x80.toByte().toInt()).toByte()

        val hex = StringBuilder()
        for (i in 0 until 16) {
            val b = hash[i].toInt() and 0xff
            if (b < 0x10) hex.append('0')
            hex.append(b.toString(16))
        }
        val formatted = "${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20, 32)}"
        return Uuid.parse(formatted)
    }
}
